# Proyecto sobre Datos socio demográficos y de recursos de salud a nivel de condado de EE. UU. (2018-2019) de Luisa

In [4]:
import pandas as pandas_lib

# Leer los datos desde la URL
url_datos = "https://raw.githubusercontent.com/4GeeksAcademy/regularized-linear-regression-project-tutorial/main/demographic_health_data.csv"
datos = pandas_lib.read_csv(url_datos)

# Mostrar las primeras filas
print(datos.head())

   fips  TOT_POP    0-9  0-9 y/o % of total pop  19-Oct  \
0  1001    55601   6787               12.206615    7637   
1  1003   218022  24757               11.355276   26913   
2  1005    24881   2732               10.980266    2960   
3  1007    22400   2456               10.964286    2596   
4  1009    57840   7095               12.266598    7570   

   10-19 y/o % of total pop  20-29  20-29 y/o % of total pop  30-39  \
0                 13.735364   6878                 12.370281   7089   
1                 12.344167  23579                 10.814964  25213   
2                 11.896628   3268                 13.134520   3201   
3                 11.589286   3029                 13.522321   3113   
4                 13.087828   6742                 11.656293   6884   

   30-39 y/o % of total pop  ...  COPD_number  diabetes_prevalence  \
0                 12.749771  ...         3644                 12.9   
1                 11.564429  ...        14692                 12.0   
2       

In [8]:
import pandas as pandas_lib

# Mostrar las primeras filas
print(data.head())

# Mostrar información del DataFrame
data.info()

# Mostrar la forma del DataFrame
print(data.shape)

# Mostrar los tipos de datos de cada columna
print(data.dtypes)

   fips  TOT_POP    0-9  0-9 y/o % of total pop  19-Oct  \
0  1001    55601   6787               12.206615    7637   
1  1003   218022  24757               11.355276   26913   
2  1005    24881   2732               10.980266    2960   
3  1007    22400   2456               10.964286    2596   
4  1009    57840   7095               12.266598    7570   

   10-19 y/o % of total pop  20-29  20-29 y/o % of total pop  30-39  \
0                 13.735364   6878                 12.370281   7089   
1                 12.344167  23579                 10.814964  25213   
2                 11.896628   3268                 13.134520   3201   
3                 11.589286   3029                 13.522321   3113   
4                 13.087828   6742                 11.656293   6884   

   30-39 y/o % of total pop  ...  COPD_number  diabetes_prevalence  \
0                 12.749771  ...         3644                 12.9   
1                 11.564429  ...        14692                 12.0   
2       

El conjunto de datos analizado contiene información demográfica y de salud de distintas regiones identificadas por el código FIPS, abarcando un total de 108 columnas y múltiples variables como población total, distribución por grupos de edad (por ejemplo, 0-9, 10-19, 20-29 años), y prevalencias de enfermedades como diabetes, EPOC (COPD) y enfermedad renal crónica (CKD), junto con sus respectivos intervalos de confianza. Las variables están representadas en distintos tipos de datos (int64 y float64) y reflejan tanto números absolutos (como número de casos) como proporciones relativas al total de la población. 

In [9]:
# Mostrar columnas de tipo object
for c in data.columns:
    if data[c].dtype == 'object':
        print(c, data[c].dtype)

COUNTY_NAME object
STATE_NAME object


In [11]:
# Codificar columnas categóricas como valores numéricos
data['COUNTY_NAME'] = pandas_lib.factorize(data['COUNTY_NAME'])[0]
data['STATE_NAME'] = pandas_lib.factorize(data['STATE_NAME'])[0]

In [12]:
# Eliminar duplicados y resetear el índice
data = data.drop_duplicates().reset_index(drop=True)

# Mostrar estadísticas descriptivas
print(data.describe())

               fips       TOT_POP           0-9  0-9 y/o % of total pop  \
count   3140.000000  3.140000e+03  3.140000e+03             3140.000000   
mean   30401.640764  1.041894e+05  1.274030e+04               11.871051   
std    15150.559265  3.335834e+05  4.180730e+04                2.124081   
min     1001.000000  8.800000e+01  0.000000e+00                0.000000   
25%    18180.500000  1.096325e+04  1.280500e+03               10.594639   
50%    29178.000000  2.580050e+04  3.057000e+03               11.802727   
75%    45081.500000  6.791300e+04  8.097000e+03               12.951840   
max    56045.000000  1.010552e+07  1.208253e+06               25.460677   

             19-Oct  10-19 y/o % of total pop         20-29  \
count  3.140000e+03               3140.000000  3.140000e+03   
mean   1.336798e+04                 12.694609  1.446933e+04   
std    4.228439e+04                  1.815044  4.957773e+04   
min    0.000000e+00                  0.000000  0.000000e+00   
25%    1.

In [13]:
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split

# Crear una lista con las columnas y quitar la variable objetivo
col = list(data.columns)
col.remove('diabetes_number')

# Definir variables predictoras y objetivo
X = data.drop('diabetes_number', axis=1)[col]
y = data['diabetes_number']

# Dividir en conjunto de entrenamiento y prueba
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


print(X_train.head())

       fips  TOT_POP    0-9  0-9 y/o % of total pop  19-Oct  \
1292  26127    26625   3221               12.097653    3463   
2302  42121    51266   5272               10.283619    5751   
761   18133    37779   3915               10.362900    5118   
2194  40131    91984  11163               12.135806   12646   
1241  26025   134487  16698               12.416070   17666   

      10-19 y/o % of total pop  20-29  20-29 y/o % of total pop  30-39  \
1292                 13.006573   2922                 10.974648   2829   
2302                 11.217961   5137                 10.020286   5341   
761                  13.547209   6202                 16.416528   4363   
2194                 13.748043  11595                 12.605453  11357   
1241                 13.135842  17281                 12.849569  15993   

      30-39 y/o % of total pop  ...  COPD_Upper 95% CI  COPD_number  \
1292                 10.625352  ...               13.0         2314   
2302                 10.418211  ..

In [ ]:
from sklearn.preprocessing import StandardScaler

# Crear el objeto scaler
scaler = StandardScaler()

# Ajustar el scaler a los datos de entrenamiento
scaler.fit(X_train)

# Transformar los datos de entrenamiento
X_train_norm = scaler.transform(X_train)
X_train_norm = pandas_lib.DataFrame(X_train_norm, index=X_train.index, columns=col)

# Transformar los datos de prueba
X_test_norm = scaler.transform(X_test)
X_test_norm = pandas_lib.DataFrame(X_test_norm, index=X_test.index, columns=col)


print(X_train_norm.head())

          fips   TOT_POP       0-9  0-9 y/o % of total pop    19-Oct  \
1292 -0.301633 -0.229763 -0.225393                0.102383 -0.231350   
2302  0.761573 -0.161280 -0.179851               -0.754597 -0.181109   
761  -0.833037 -0.198764 -0.209983               -0.717144 -0.195009   
2194  0.629287 -0.048115 -0.049041                0.120407 -0.029705   
1241 -0.308413  0.070012  0.073864                0.252809  0.080526   

      10-19 y/o % of total pop     20-29  20-29 y/o % of total pop     30-39  \
1292                  0.162374 -0.229775                 -0.429454 -0.223780   
2302                 -0.836073 -0.188375                 -0.736296 -0.176225   
761                   0.464170 -0.168470                  1.320194 -0.194740   
2194                  0.576280 -0.067671                  0.094875 -0.062335   
1241                  0.234535  0.038603                  0.173362  0.025430   

      30-39 y/o % of total pop  ...  COPD_Upper 95% CI  COPD_number  \
1292           

In [16]:
import os

# Crear el directorio si no existe
output_dir = "/workspaces/regularizedLinearReg/data/processed"
os.makedirs(output_dir, exist_ok=True)

# Agregar la variable objetivo a los conjuntos de entrenamiento y prueba normalizados
X_train_norm["diabetes_number"] = list(y_train)
X_test_norm["diabetes_number"] = list(y_test)

# Guardar los conjuntos de datos normalizados en archivos CSV
X_train_norm.to_csv(os.path.join(output_dir, "data_train.csv"), index=False)
X_test_norm.to_csv(os.path.join(output_dir, "data_test.csv"), index=False)

In [ ]:
# Leer los datos de entrenamiento y prueba desde los archivos CSV
train_data = pandas_lib.read_csv("/workspaces/regularizedLinearReg/data/processed/data_train.csv")
test_data = pandas_lib.read_csv("/workspaces/regularizedLinearReg/data/processed/data_test.csv")


print(train_data.head())

       fips   TOT_POP       0-9  0-9 y/o % of total pop    19-Oct  \
0 -0.301633 -0.229763 -0.225393                0.102383 -0.231350   
1  0.761573 -0.161280 -0.179851               -0.754597 -0.181109   
2 -0.833037 -0.198764 -0.209983               -0.717144 -0.195009   
3  0.629287 -0.048115 -0.049041                0.120407 -0.029705   
4 -0.308413  0.070012  0.073864                0.252809  0.080526   

   10-19 y/o % of total pop     20-29  20-29 y/o % of total pop     30-39  \
0                  0.162374 -0.229775                 -0.429454 -0.223780   
1                 -0.836073 -0.188375                 -0.736296 -0.176225   
2                  0.464170 -0.168470                  1.320194 -0.194740   
3                  0.576280 -0.067671                  0.094875 -0.062335   
4                  0.234535  0.038603                  0.173362  0.025430   

   30-39 y/o % of total pop  ...  COPD_number  diabetes_prevalence  \
0                 -0.665485  ...    -0.222477       

In [18]:
# Separar las características (X) y la variable objetivo (y) para los datos de entrenamiento y prueba
X_train = train_data.drop(["diabetes_number"], axis=1)
y_train = train_data["diabetes_number"]
X_test = test_data.drop(["diabetes_number"], axis=1)
y_test = test_data["diabetes_number"]

In [ ]:
from sklearn.feature_selection import SelectKBest, f_regression

# Seleccionar el 50% de las características
k = int(len(X_train.columns) * 0.5)

# Crear el modelo de selección
selection_model = SelectKBest(score_func=f_regression, k=k)

# Ajustar el modelo a los datos de entrenamiento
selection_model.fit(X_train, y_train)

# Obtener las características seleccionadas
ix = selection_model.get_support()

# Transformar los datos de entrenamiento y prueba según las características seleccionadas
X_train_sel = pandas_lib.DataFrame(selection_model.transform(X_train), columns = X_train.columns.values[ix])
X_test_sel = pandas_lib.DataFrame(selection_model.transform(X_test), columns = X_test.columns.values[ix])


print(X_train_sel.head())


    TOT_POP       0-9    19-Oct     20-29     30-39  30-39 y/o % of total pop  \
0 -0.229763 -0.225393 -0.231350 -0.229775 -0.223780                 -0.665485   
1 -0.161280 -0.179851 -0.181109 -0.188375 -0.176225                 -0.785934   
2 -0.198764 -0.209983 -0.195009 -0.168470 -0.194740                 -0.128551   
3 -0.048115 -0.049041 -0.029705 -0.067671 -0.062335                  0.335452   
4  0.070012  0.073864  0.080526  0.038603  0.025430                  0.070962   

      40-49     50-59     60-69     70-79  ...  anycondition_number  \
0 -0.228216 -0.227030 -0.231375 -0.232151  ...            -0.231574   
1 -0.163303 -0.139120 -0.112664 -0.124916  ...            -0.146243   
2 -0.193726 -0.199360 -0.218215 -0.220562  ...            -0.193279   
3 -0.044847 -0.033104 -0.044844 -0.036279  ...            -0.016632   
4  0.048850  0.079829  0.119807  0.116625  ...             0.140471   

   Obesity_Upper 95% CI  Obesity_number  Heart disease_prevalence  \
0              0.

In [20]:
from sklearn.linear_model import LinearRegression

# Crear el modelo de regresión lineal
model = LinearRegression()

# Ajustar el modelo a los datos de entrenamiento
model.fit(X_train_sel, y_train)


LinearRegression()

In [21]:
from sklearn.metrics import mean_squared_error, r2_score

# Imprimir el intercepto y los coeficientes del modelo
print(f"Intercep (a): {model.intercept_}")
print(f"Coeficientes: {model.coef_}")


Intercep (a): 9762.924363057327
Coeficientes: [-6.33214372e+03  8.54376295e+03 -1.14085937e+04 -2.43615206e+04
 -8.28318662e+03  7.10013055e+01 -1.28642403e+03 -1.59875127e+03
 -4.97484287e+03 -5.10425265e+03  2.13548401e+03 -8.74163069e+03
 -2.65601121e+02 -3.53169031e+02 -3.87760393e+02 -2.27864115e+01
  1.07009176e+02 -4.70061729e+02 -6.33214372e+03 -1.42446951e+03
 -9.39124403e+02  3.91925734e+01 -7.14537010e+03 -1.17042409e+04
 -2.32368066e+04 -2.68692329e+04 -2.28760729e+02 -6.84774615e+01
  3.00127026e+03  1.26270341e+02 -5.34277395e+03 -5.58461846e+03
  3.42016651e+02  1.22461511e+03 -4.83657406e+02 -3.06970178e+02
 -7.76316930e+03 -1.42019517e+02  6.41414590e+03  4.20148666e+02
 -4.73143115e+03 -5.00501488e+03  1.70045235e+05  2.54179678e+04
 -1.44437124e+02 -6.70544686e+03  1.46080617e+02  6.22181712e+02
 -5.52279209e+02  8.65061741e+03 -8.04199992e+03 -2.34579248e+03
 -1.39183409e+01]


In [22]:
# Realizar las predicciones sobre el conjunto de prueba
y_pred = model.predict(X_test_sel)


print(y_pred)


[ 1.07106265e+03  2.08416378e+04  2.65412596e+03  9.26287208e+03
  1.71793480e+04  2.45170152e+03  9.61715573e+02  1.53552960e+03
  2.25468996e+02  1.57679393e+03  1.61177414e+03  4.84622027e+02
  6.77024123e+03  1.87404146e+03  1.37219181e+04  4.02535231e+03
  9.18719689e+04  2.01430448e+03  3.68484210e+03  5.23068031e+02
  2.56808272e+03  6.25725162e+02  1.00092505e+03  7.53269857e+02
  1.88994504e+05  5.10519422e+03  1.89797147e+03  2.63309603e+03
  9.26721881e+02  1.78098750e+02  1.76645603e+03  3.67080553e+03
  3.71394310e+03  4.49002599e+03  1.21364035e+04  2.22505464e+03
  4.71178459e+03  1.70330490e+03  9.69755457e+02  2.82553128e+03
  2.51180955e+03  1.51112104e+04  3.44772004e+02  3.74129075e+02
  4.37643467e+03  7.62808963e+02  2.99510834e+02  1.15320489e+04
  1.60445732e+03  9.87635553e+02  2.55775872e+04  6.86462072e+02
  1.95308380e+03  1.11763425e+03  1.58034904e+04  1.61226946e+03
  5.42169845e+04  2.45079377e+03  4.60694422e+03  9.72056092e+02
  1.21449852e+04  5.19773

In [23]:
# Calcular el score (R2) en el conjunto de prueba
base_score = model.score(X_test_sel, y_test)

print("Coeficientes:", model.coef_)
print("R2 score:", base_score)


Coeficientes: [-6.33214372e+03  8.54376295e+03 -1.14085937e+04 -2.43615206e+04
 -8.28318662e+03  7.10013055e+01 -1.28642403e+03 -1.59875127e+03
 -4.97484287e+03 -5.10425265e+03  2.13548401e+03 -8.74163069e+03
 -2.65601121e+02 -3.53169031e+02 -3.87760393e+02 -2.27864115e+01
  1.07009176e+02 -4.70061729e+02 -6.33214372e+03 -1.42446951e+03
 -9.39124403e+02  3.91925734e+01 -7.14537010e+03 -1.17042409e+04
 -2.32368066e+04 -2.68692329e+04 -2.28760729e+02 -6.84774615e+01
  3.00127026e+03  1.26270341e+02 -5.34277395e+03 -5.58461846e+03
  3.42016651e+02  1.22461511e+03 -4.83657406e+02 -3.06970178e+02
 -7.76316930e+03 -1.42019517e+02  6.41414590e+03  4.20148666e+02
 -4.73143115e+03 -5.00501488e+03  1.70045235e+05  2.54179678e+04
 -1.44437124e+02 -6.70544686e+03  1.46080617e+02  6.22181712e+02
 -5.52279209e+02  8.65061741e+03 -8.04199992e+03 -2.34579248e+03
 -1.39183409e+01]
R2 score: 0.9952738787827744


El modelo de regresión lineal presenta un excelente desempeño, con un R2 score de 0.995, lo que indica que el 99.53% de la variabilidad en la variable dependiente (diabetes_number) se explica por las características del modelo. Los coeficientes muestran una relación positiva o negativa con las variables predictoras, lo que sugiere cómo cada característica influye en el número de casos de diabetes. 

In [24]:
from sklearn.linear_model import Lasso

# Cargar los datos de entrenamiento y prueba (deben estar normalizados y procesados previamente)
lasso_model = Lasso()

# Ajustar el modelo Lasso a los datos de entrenamiento
lasso_model.fit(X_train_sel, y_train)


/home/vscode/.local/lib/python3.11/site-packages/sklearn/linear_model/_coordinate_descent.py:695: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 1.816e+09, tolerance: 2.595e+08
  model = cd_fast.enet_coordinate_descent(


Lasso()

In [25]:
# Calcular el score (R2) del modelo Lasso en el conjunto de prueba
score = lasso_model.score(X_test_sel, y_test)


print("Coeficientes:", lasso_model.coef_)
print("R2 score:", score)


Coeficientes: [ 2.06004484e+04 -5.77178780e+03 -1.36346071e+03  4.97404351e+01
  1.68125686e+02  3.15540858e+01  1.37757273e+04  4.98366166e+03
 -9.24955996e+02 -4.03126933e+02  2.45240450e+03 -1.09132716e+04
 -2.05584306e+02 -4.57867184e+02 -8.90905874e+02 -4.76307076e+01
  2.25592983e+02 -9.38298285e+02 -3.15993556e-01 -7.38256627e+01
 -1.32303927e+02  1.37257507e+02  4.90940307e+03  3.74325317e+03
 -3.76573512e+03 -5.34522759e+03 -2.11092087e+02  1.45737788e+02
  4.93961135e+03  2.41481550e+01 -5.04001470e+03 -0.00000000e+00
 -5.49136951e+02  9.29476746e+02 -5.27779364e+02 -8.73920529e+01
 -9.37585634e+03 -6.27451776e+02  8.07212449e+03 -0.00000000e+00
  1.44980952e+03  8.98694266e+03 -2.96938166e+03  1.81247602e+04
 -1.08020127e+02 -3.99786537e+03 -0.00000000e+00  7.93991837e+02
 -5.16609512e+02  7.01944843e+03 -9.12513976e+03 -5.54165145e+03
 -3.14946359e+01]
R2 score: 0.9948715604006845


El modelo Lasso muestra un buen desempeño con un R2 score de 0.995, lo que indica que el 99.49% de la variabilidad en la variable dependiente (diabetes_number) es explicada por las características seleccionadas. Los coeficientes muestran que algunas variables tienen un impacto significativo, con coeficientes altos y bajos, tanto positivos como negativos, lo que sugiere que el aumento o disminución de ciertas características puede influir en los casos de diabetes. Además, algunos coeficientes se acercan a cero, lo que indica que estas variables no tienen una relación significativa con la variable objetivo, lo que es típico en modelos Lasso que realizan selección de características. Este alto R2 score sugiere un buen ajuste del modelo, aunque sería prudente validar el modelo en diferentes conjuntos de datos para asegurarse de que no haya sobreajuste.

In [26]:
from sklearn.linear_model import Lasso

# Ajustar el modelo Lasso con parámetros específicos
lasso_model = Lasso(alpha=0.1, max_iter=100)

# Ajustar el modelo Lasso a los datos de entrenamiento
lasso_model.fit(X_train_sel, y_train)

/home/vscode/.local/lib/python3.11/site-packages/sklearn/linear_model/_coordinate_descent.py:695: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 3.135e+09, tolerance: 2.595e+08
  model = cd_fast.enet_coordinate_descent(


Lasso(alpha=0.1, max_iter=100)

In [27]:
# Calcular el score (R2) del modelo Lasso en el conjunto de prueba
score = lasso_model.score(X_test_sel, y_test)


print("Coeficientes:", lasso_model.coef_)
print("R2 score:", score)


Coeficientes: [ 2.81734268e+04 -2.77992124e+03  3.05295513e+02 -1.27953968e+03
  1.27366037e+03  7.24150205e+01  3.86215408e+03  2.15712477e+03
 -2.08268771e+01  1.76221665e+03  1.66524076e+03 -7.03406471e+03
  5.11219820e+02 -6.77369342e+02  4.66946622e+02 -1.52632838e+02
  7.87989552e+02 -1.96950720e+03  1.24335733e+03 -3.02877411e+02
 -4.28747632e+01  1.84337851e+02  4.29006620e+03  2.79976726e+03
 -4.42395539e+03 -6.87833490e+03 -1.66110819e+02  2.80198258e+02
  4.23520356e+03  1.87508910e+02 -4.92177303e+02 -1.05459101e+01
  2.34230956e+02  1.47997117e+03 -1.08489859e+02 -1.95425786e+02
 -4.92142442e+03  5.61749457e+02  1.84526843e+03 -1.02980712e+02
  1.42208344e+03  4.65887883e+03 -3.65198167e+03  4.77681207e+03
  6.24392566e+01 -1.35311728e+03  6.15283299e+02  4.71201447e+02
 -5.91230172e+02  3.46195725e+03 -2.52155405e+03 -1.39588457e+03
 -9.35726723e+01]
R2 score: 0.9910816672523963


In [28]:
from sklearn.linear_model import Lasso

# Ajustar el modelo Lasso con un alpha de 0.4 y 150 iteraciones
lasso_model = Lasso(alpha=0.4, max_iter=150)

# Ajustar el modelo Lasso a los datos de entrenamiento
lasso_model.fit(X_train_sel, y_train)


/home/vscode/.local/lib/python3.11/site-packages/sklearn/linear_model/_coordinate_descent.py:695: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 2.921e+09, tolerance: 2.595e+08
  model = cd_fast.enet_coordinate_descent(


Lasso(alpha=0.4, max_iter=150)

In [29]:
# Calcular el score (R2) del modelo Lasso en el conjunto de prueba
score = lasso_model.score(X_test_sel, y_test)

print("Coeficientes:", lasso_model.coef_)
print("R2 score:", score)


Coeficientes: [ 2.75102502e+04 -3.48572311e+03  2.93352508e+02 -1.35769562e+03
  1.69274834e+03  5.17083888e+01  4.96287799e+03  2.54757392e+03
 -3.43917913e+02  1.55102213e+03  1.98298150e+03 -7.79867689e+03
  4.11258268e+02 -6.72279102e+02  2.61432511e+02 -1.36082730e+02
  6.20884487e+02 -1.68370938e+03  1.17947328e+03 -2.48482810e+02
  1.30913416e+01  1.79193691e+02  4.48944874e+03  3.09366374e+03
 -4.45948057e+03 -7.23066112e+03 -1.87124019e+02  2.69321044e+02
  3.90282272e+03  1.58941730e+02 -3.09652237e+02 -9.52733266e+00
  1.10592397e+02  1.18986991e+03 -1.18615375e+02  1.52270836e+01
 -5.52765209e+03  2.88219168e+02  2.85542681e+03 -3.59331761e+01
  1.34119583e+03  5.33365254e+03 -3.90715647e+03  6.16140294e+03
  3.38723571e+01 -1.67959580e+03  5.60957721e+02  6.69616279e+02
 -7.51878317e+02  3.66721294e+03 -3.40092999e+03 -1.47451622e+03
 -8.16896878e+01]
R2 score: 0.9918617572301807


Conclusión: Teniendo en cuenta la modificación de los hiperparámetros, se observa que, aunque el aumento en el valor del alpha ha llevado a un leve incremento en el R2, el segundo modelo sigue siendo más robusto que el primero. Ambos modelos lograron un rendimiento sobresaliente con un R2 superior al 99% en los datos de prueba, lo que indica que son capaces de hacer predicciones muy precisas. Sin embargo, el modelo con un alpha mayor muestra una ligera mejora en la capacidad de regularización, aunque no afecta significativamente el rendimiento general.